src/components/DropDownController.jsx

```diff
@@ -0,0 +1,501 @@
+          openDirection === OPEN_UP && styles.DropDownController_picker__openUp,
+          !withAnyPortal && openDirection === OPEN_DOWN && {
+            top: inputHeight + verticalSpacing,
+          },
+          !withAnyPortal && openDirection === OPEN_UP && {
+            bottom: inputHeight + verticalSpacing,
+          },
+          orientation === HORIZONTAL_ORIENTATION && styles.DropDownController_picker__horizontal,
+          orientation === VERTICAL_ORIENTATION && styles.DropDownController_picker__vertical,
+          withAnyPortal && styles.DropDownController_picker__portal,
+          withFullScreenPortal && styles.DropDownController_picker__fullScreenPortal,
+          isRTL && styles.DropDownController_picker__rtl,
+          dayPickerContainerStyles,
+        )}
+        onClick={onOutsideClick}
+      >
+        {DropDown && (
+          <DropDown
+            {...dropdownProps}
+            isFocused={isDayPickerFocused}
+            showKeyboardShortcuts={showKeyboardShortcuts}
+            onBlur={this.onDayPickerBlur}
+          />
+        )}
+
+        {withFullScreenPortal && (
+          <button
+            {...css(styles.DropDownController_closeButton)}
+            aria-label={(phrases || {}).closeDatePicker}
+            type="button"
+            onClick={this.onOutsideClick}
+          >
+            <div {...css(styles.DropDownController_closeButton_svg)}>
+              {closeIcon}
+            </div>
+          </button>
+        )}
+      </div>
+    );
+  }
+
+  render() {
+    const {
+      withFullScreenPortal,
+      withPortal,
+      verticalSpacing,
+      block,
+      input: Input,
+      inputProps,
+      styles,
+    } = this.props;
+
+    const { isInputFocused } = this.state;
+
+    const enableOutsideClick = (!withPortal && !withFullScreenPortal);
+    const hideFang = verticalSpacing < FANG_HEIGHT_PX;
+
+    const inputEl = Input && (
+      <Input
+        {...inputProps}
+        showCaret={!withPortal && !withFullScreenPortal && !hideFang}
+        isFocused={isInputFocused}
+        onFocusChange={this.onInputFocus}
+        onKeyDownArrowDown={this.onDayPickerFocus}
+        onKeyDownQuestionMark={this.showKeyboardShortcutsPanel}
+      />
+    );
+
+    return (
+      <div
+        ref={this.setContainerRef}
+        {...css(
+          styles.DropDownController,
+          block && styles.DropDownController__block,
+        )}
+      >
+        {enableOutsideClick && (
+          <OutsideClickHandler onOutsideClick={this.onOutsideClick}>
+            {inputEl}
+            {this.maybeRenderDayPickerWithPortal()}
+          </OutsideClickHandler>
+        )}
+        {!enableOutsideClick && inputEl}
+        {!enableOutsideClick && this.maybeRenderDayPickerWithPortal()}
+      </div>
+    );
+  }
+}
+
+DropDownController.propTypes = propTypes;
+DropDownController.defaultProps = defaultProps;
+
+export { DropDownController as PureDropDownController };
```

Comment:
this is an antipattern we should avoid; can `.dive()` be used instead?

Reply:
A lot of this file was simply pulled out from the `SingleDatePicker`. The reason for this pattern is the same as it was previously there---which is that we're spying on and stubbing instance methods. Unfortunately, dive does not have a good story for that case.